engine/gamesys/src/gamesys/scripts/script_box2d.cpp

```diff
@@ -0,0 +1,190 @@
+#include <Box2D/Dynamics/Joints/b2Joint.h>
+
+#include <dlib/log.h>
+#include <gameobject/script.h>
+
+#include "gamesys.h"
+#include "../gamesys_private.h"
+
+#include "components/comp_collision_object.h"
+
+#include "script_box2d.h"
+
+extern "C"
+{
+#include <lua/lauxlib.h>
+#include <lua/lualib.h>
+}
+
+namespace dmGameSystem
+{
+    static float g_PhysicsScale = 1.0f;
+    static float g_InvPhysicsScale = 1.0f / g_PhysicsScale;
+
+    //////////////////////////////////////////////////////////////////////////////
+    // Types
+
+    enum Box2DUserType
+    {
+        BOX2D_TYPE_BODY,
+        BOX2D_TYPE_MAX,
+    };
+
+    static uint32_t TYPE_HASHES[BOX2D_TYPE_MAX];
+
+    #define BOX2D_TYPE_NAME_BODY "b2body"
+
+    //////////////////////////////////////////////////////////////////////////////
+    // b2Vec2
+
+    b2Vec2 CheckVec2(lua_State* L, int index, float scale)
+    {
+        dmVMath::Vector3* v = dmScript::CheckVector3(L, index);
+        return b2Vec2(v->getX() * scale, v->getY() * scale);
+    }
+
+    dmVMath::Vector3 FromB2(const b2Vec2& p, float inv_scale)
+    {
+        return dmVMath::Vector3(p.x * inv_scale, p.y * inv_scale, 0);
+    }
+
+    void PushWorld(struct lua_State* L, class b2World* world)
+    {
+        lua_pushlightuserdata(L, world);
+    }
+
+    void SetPhysicsScale(float scale)
+    {
+        g_PhysicsScale = scale;
+        g_InvPhysicsScale = 1.0f / g_PhysicsScale;
+    }
+
+    float GetPhysicsScale()
+    {
+        return g_PhysicsScale;
+    }
+
+    float GetInvPhysicsScale()
+    {
+        return g_InvPhysicsScale;
+    }
+
+    //////////////////////////////////////////////////////////////////////////////
+
+    static void GetCollisionObject(lua_State* L, int index, dmGameObject::HCollection collection, void** comp, void** comp_world)
+    {
+        dmGameObject::GetComponentUserDataFromLua(L, index, collection, COLLISION_OBJECT_EXT, (uintptr_t*)comp, 0, comp_world);
+    }
+
+    static int B2D_GetWorld(lua_State* L)
+    {
+        DM_LUA_STACK_CHECK(L, 1);
+
+        dmGameObject::HCollection collection = dmGameObject::GetCollection(CheckGoInstance(L));
+        uint32_t component_type_index = dmGameObject::GetComponentTypeIndex(collection, COLLISION_OBJECT_EXT_HASH);
+        void* comp_world = dmGameObject::GetWorld(collection, component_type_index);
+        b2World* world = dmGameSystem::CompCollisionObjectGetBox2DWorld(comp_world);
+
+        if (world)
+            PushWorld(L, world);
+        else
+            lua_pushnil(L);
+        return 1;
+    }
+
+    static int B2D_GetBody(lua_State* L)
+    {
+        DM_LUA_STACK_CHECK(L, 1);
+
+        dmGameObject::HCollection collection = dmGameObject::GetCollection(CheckGoInstance(L));
+
+        void* component = 0;
+        GetCollisionObject(L, 1, collection, &component, 0);
+
+        b2Body* body = dmGameSystem::CompCollisionObjectGetBox2DBody(component);
+
+        if (body)
+            PushBody(L, body);
```

Comment:
I need to make this more robust, as it currently contains an internal pointer, and the game object may be deleted directly after (and along with is the collision object).

E.g. if this has a game object id set, then we need to check it before doing any operations on the body.